modules/select2_facets/src/Controller/FacetApiAutocompleteController.php

```diff
@@ -46,31 +36,14 @@ class FacetApiAutocompleteController extends ControllerBase {
    */
   protected $storedRequests = [];
 
-  /**
-   * Constructs a FacetApiAutocompleteController object.
-   *
-   * @param \Drupal\Core\KeyValueStore\KeyValueStoreInterface $key_value
-   *   The key value factory.
-   * @param \Drupal\facets\FacetManager\DefaultFacetManager $facetManager
-   *   The facet manager service.
-   * @param \Symfony\Component\HttpFoundation\RequestStack $requestStack
-   *   The request stack.
-   */
-  public function __construct(KeyValueStoreInterface $key_value, DefaultFacetManager $facetManager, RequestStack $requestStack) {
-    $this->keyValue = $key_value;
-    $this->facetManager = $facetManager;
-    $this->requestStack = $requestStack;
-  }
-
   /**
    * {@inheritdoc}
    */
   public static function create(ContainerInterface $container) {
-    return new static(
-      $container->get('keyvalue')->get('entity_autocomplete'),
-      $container->get('facets.manager'),
-      $container->get('request_stack')
-    );
+    $controller = parent::create($container);
+    $controller->facetManager = $container->get('facets.manager');
```

Comment:
I might consider an assert against the expected interface - this keeps the benefits of the constructor typehinting without downsides of the overriding the constructor - the other thing to consider are protected setters - you can call them here - so new api but you get typehinting.

Reply:
And this is the same for the other ::create() overrides. Add typehinting in some form and this looks great. The module will be isolated from constructor changes. 

There is one thought though. All the things you are extending are meant to be extended ie. ControllerBase and WidgetPluginBase and if core broken these constructors then we'd be doing it wrong too. So there is an argument that these particular changes are unnecessary. I think I should be clearer and say that my advice is for when you extend an concrete and non-base class plugin / controller / form whatever from core.

Reply:
@alexpott What about concrete classes, that do not implement the ContainerFactoryPluginInterface? Neither they themselves nor their base.
Are they guaranteed to be stable or just a future problem we cannot circumvent?

Reply:
I think plugin base classes should implement ContainerFactoryPluginInterface. For example, if Drupal\Core\Field\WidgetBase would implement ContainerFactoryPluginInterface that would make all the child classes more BC stable.

Reply:
oh good point! That is very sensible.